docs/Cursors.md

```diff
@@ -0,0 +1,287 @@
+
+- `p.args()`: Returns cursors to the procedure's arguments.
+- `p.body()`: Returns a `BlockCursor` selecting the entire body of the procedure.
+- `p.find(pattern, many=False)`: Finds cursor(s) matching the given `pattern` string:
+   - If `many=False` (default), returns the first matching cursor. 
+   - If `many=True`, returns a list of all matching cursors.
+- `p.find_loop(loop_pattern, many=False)`: Finds cursor(s) to a loop, expanding shorthand patterns:
+   - `"name"` or `"name #n"` are expanded to `for name in _:_`
+   - Works like `p.find()`, returning the first match by default unless `many=True`
+- `p.find_alloc_or_arg(buf_name)`: Finds an allocation or argument cursor, expanding the name to `buf_name: _`.
+- `p.find_all(pattern)`: Shorthand for `p.find(pattern, many=True)`, returning all matching cursors.
+
+### From Cursors
+A `Cursor` provides a similar method to find sub-cursors within its sub-AST:
+
+- `c.find(pattern, many=False)`: Finds cursor(s) matching the `pattern` within the cursor's sub-AST.
+  - Like `p.find()`, returns the first match by default unless `many=True`.
+
+### Pattern Language
+The `pattern` argument is a string using the following special syntax:
+
+- `_` is a wildcard matching any statement or expression
+- `#n` at the end selects the `n+1`th match instead of the first 
+  - Ex. `"for i in _:_ #2"` matches the 3rd `i` loop
+- `;` is a sequence of statements
+
+Example patterns:
+- `"for i in _:_"` matches a `for i in _:_` loop
+- `"if i == 0:_"` or `"if _:_"` match `if` statements 
+- `"a : i8"` or `"a : _"` match an allocation of buffer `a`
+- `"a = 3.0"` or `"a = _"` match an assignment to `a`
+- `"a += 3.0"` or `"a += _"` match a reduction on `a`
+- `"a = 3.0 ; b = 2.0"` matches a block with those two statements
+
+## Cursor Types
+
+Exo defines the following `Cursor` types:
+
+- `StmtCursor`: Cursor to a specific Exo IR statement
+- `GapCursor`: Cursor to the space between statements, anchored to a statement
+- `BlockCursor`: Cursor to a block (sequence) of statements
+- `ArgCursor`: Cursor to a procedure argument (no navigation)
+- `InvalidCursor`: Special cursor type for invalid cursors
+
+## Common Cursor Methods
+
+All `Cursor` types provide these common methods:
+
+- `c.parents()`: Returns cursor to parent node in Exo IR
+  - Raises `InvalidCursorError` if at the root with no parent
+- `c.proc()`: Returns the `Procedure` this cursor is associated with
+- `c.find(pattern, many=False)`: Finds matches within cursor's sub-AST
+
+## Statement Cursor Navigation
+
+A `StmtCursor` (pointing to one IR statement) provides these navigation methods:
+
+- `c.next()`: Returns cursor to next statement 
+- `c.prev()`: Returns cursor to previous statement
+- `c.before()`: Returns `GapCursor` to space immediately before this statement
+- `c.after()`: Returns `GapCursor` to space immediately after this statement 
+- `c.as_block()`: Returns a `BlockCursor` containing only this one statement
+
```

Comment:
I would also advertise the `StmtCursor.expand()` shorthand here. It's actually quite useful for correctly handling the else case (when learning cursors I didn't know about expand and hacked together `StmtCursor.parent().body()` which is incorrect when the StmtCursor points inside an orelse).